torch/csrc/jit/codegen/cuda/lower_shift.cpp

```diff
@@ -0,0 +1,684 @@
+  const auto shift_pred = predicates[0];
+  const auto padding_pred = predicates[1];
+
+  // If null, no specific predicate is needed.
+  if (shift_pred == nullptr) {
+    return;
+  }
+
+  // The conditional branches to create:
+  //
+  // if (shift_pred) {
+  //   consumer = producer;
+  // } else {
+  //   if (padding_pred) {
+  //     consumer = 0;
+  //   }
+  // }
+
+  auto shift_ite = ir_builder.create<kir::IfThenElse>(shift_pred);
+
+  auto& scope = loops.back()->body();
+
+  // Insert the if statement
+  scope.insert_before(expr, shift_ite);
+
+  // Remove the expr from the list
+  scope.erase(expr);
+
+  // Place the expr inside the if statement
+  shift_ite->thenBody().push_back(expr);
+
+  // Pading by zero
+  auto bounds_ite = ir_builder.create<kir::IfThenElse>(padding_pred);
+  const int pad_value = 0;
+  auto pad_expr = ir_builder.create<kir::UnaryOp>(
+      UnaryOpType::Set, out_tv, ir_builder.create<kir::Int>(pad_value));
+  bounds_ite->thenBody().push_back(pad_expr);
+  // Insert the else block
+  shift_ite->elseBody().push_back(bounds_ite);
+}
+
+std::array<kir::Bool*, 2> ShiftPredicateInserter::getPredicate(
+    const kir::Expr* expr,
+    const std::vector<kir::ForLoop*>& loops,
+    kir::TensorView* out_tv) {
+  const auto gpu_lower = GpuLower::current();
+  kir::IrBuilder ir_builder(gpu_lower->kernel());
+
+  TensorView* out_fuser_tv = out_tv->fuserTv();
+
+  const bool needs_shift_predicate =
+      gpu_lower->haloInfo().needsShiftPredicate(out_fuser_tv->definition());
+
+  if (!needs_shift_predicate) {
+    return {nullptr, nullptr};
+  }
+
+  const auto& root_domain = out_fuser_tv->getRootDomain();
+
+  auto shift_expr = dynamic_cast<ShiftOp*>(out_fuser_tv->definition());
+
+  // Creates indices at the root domain.
+  // Set contiguity of all axes false as separate indices are needed for each
+  // root axis
+  const auto pred_contiguity = std::vector<bool>(root_domain.size(), false);
+  auto indices =
+      Index::getConsumerRootPredIndices(out_tv, loops, pred_contiguity).first;
+  TORCH_INTERNAL_ASSERT(indices.size() == root_domain.size());
+
+  kir::Bool* shift_pred = nullptr;
+  kir::Bool* padding_pred = nullptr;
+
+  for (size_t i = 0; i < root_domain.size(); ++i) {
+    auto root_id = root_domain[i];
+
+    const auto halo_info = gpu_lower->haloInfo().getRootAxisInfo(root_id);
+
+    const int shift_offset =
+        (shift_expr != nullptr) ? shift_expr->offset(i) : 0;
+
+    // "left" means halo at offset zero.
+    // shifted accesses when idx >= left_limit. padding if idx <
+    // left_limit.
+
+    // The elements at the left halo region are just set by the
+    // padding value.
+    unsigned left_limit = halo_info.width(0);
+
+    // If the defining expr is ShiftOp and its offset is positive,
+    // consumer access at 0 to the offset corresponds to out-of-bound
+    // producer access
+    if (shift_offset > 0) {
+      left_limit += (unsigned)shift_offset;
+    }
```

Comment:
I think this is correct after reviewing but need confirmation.

Reply:
Yes, that's right. 

This is actually in some cases redundant. This is from the `Shift3ptStencil_CUDA` test:

```
  for(size_t ki54 = 0; ki54 < (ceilDiv(T0.size[0], 4)); ++ki54) {
    float T6[(4 + 2)];
    #pragma unroll 1
    for(size_t ki69 = 0; ki69 < (4 + 2); ++ki69) {
      if (((((ki54 * 4) + ki69) >= 1) && (((ki54 * 4) + ki69) < (T0.size[0] + 1)))) {
        T6[ki69]
           = T0[((((ki54 * 4) + ki69) + -1) * T0.stride[0])];
      } else {
        if ((((ki54 * 4) + ki69) < (T0.size[0] + 2))) {
          T6[ki69] = 0;
        }
      }
    }
    #pragma unroll 1
    for(size_t ki56 = 0; ki56 < 4; ++ki56) {
      float T2[1];
      if (((((ki54 * 4) + ki56) >= 1) && (((ki54 * 4) + ki56) < T0.size[0]))) {
        T2[0]
           = T6[ki56];
      } else {
        if ((((ki54 * 4) + ki56) < T0.size[0])) {
          T2[0] = 0;
        }
      }
```

Here, `T2` is a 1D tensor defined as `shift(T6, {1})`. `T2` itself doesn't need to have halo. However, because of the positive shift offset, the `left_limit` value becomes 1, and that's why the predicate has `((ki54 * 4) + ki56) >= 1`. When it's false, i.e., when indexing into offset 0 of `T2`, I chose to always take the padding path since the corresponding producer point is out of bound. However, when the producer is also extended with halo (like `T6` in this case), the out-of-bound producer point actually corresponds to a halo point, so this predication is actually redundant in such a case.

However, a producer may not have halo (for example, input tensors don't have halo), and in that case, this predication is needed. 

I'm going to add a note in the code that this may be redundant and should be revisited for optimization.

Reply:
Thank you for the explanation.